### PR TITLE
[Torch] Add general AdaptiveAvgPool2dOp decompose support

### PR DESCRIPTION
**Pull request by Xinyu302**
Previously, it could only handle the situations where outputsize == (1, 1) or outputsize == (input_H, input_W). Now it supports all situations where input_H % output_H== 0 && input_W % output_W == 0

### CODE REVIEW

**Review by zjgarvey — CHANGES_REQUESTED** [review]
Thanks, this is a good addition to the decomposition. Do you plan on adding the same support for the 1d decomposition?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -5626,19 +5627,28 @@ class DecomposeAtenAdaptiveAvgPool2dOp
...
+        Value stride = rewriter.create<AtenFloordivIntOp>(
+            loc, inputHW[i], outputShapeSizesTorchInt[i]);
         Value outMinusOne = rewriter.create<AtenSubIntOp>(
             loc, outputShapeSizesTorchInt[i], constantOne);
-        kernelSize.push_back(
-            rewriter.create<AtenSubIntOp>(loc, inputHW[i], outMinusOne));
+        Value kernelSizeValue = rewriter.create<AtenSubIntOp>(
+            loc, inputHW[i],
+            rewriter.create<AtenMulIntOp>(loc, outMinusOne, stride));
+        kernelSize.push_back(kernelSizeValue);
```
[the remark this answers is not in the document]
Comment (Xinyu302) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:5650]: Yes, I think the simplification can be applied

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 2]
```diff
@@ -54,6 +54,49 @@ def AdaptiveAvgPool2dNonUnitOutputSizeDynamicModule_basic(
         module, tu: TestUtils):
     module.forward(tu.rand(1, 512, 7, 7))
 
+
+class AdaptiveAvgPool2dGeneralOutputSizeDynamicModule(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (Xinyu302) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py]: Change into AdaptiveAvgPool2dOutputSizeDivisibleByInputStaticModule

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 3]
```diff
@@ -5650,8 +5660,7 @@ class DecomposeAtenAdaptiveAvgPool2dOp
     // size is same as the input size. Therfore, keeping the stride as one for
     // the latter case as well for the ease of implementation.
     Value strideList = rewriter.create<PrimListConstructOp>(
-        loc, Torch::ListType::get(Torch::IntType::get(context)),
-        ValueRange{constantOne, constantOne});
+        loc, Torch::ListType::get(Torch::IntType::get(context)), strides);
```
[the remark this answers is not in the document]
Comment (Xinyu302) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:5653]: I have deleted the comment.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 4]
```diff
@@ -54,6 +54,49 @@ def AdaptiveAvgPool2dNonUnitOutputSizeDynamicModule_basic(
...
+    
+    def __init__(self):
+        super().__init__()
+        self.aap2d = torch.nn.AdaptiveAvgPool2d((7, 7))
```
[the remark this answers is not in the document]
Comment (Xinyu302) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py]: fixed

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 5]
```diff
@@ -5626,32 +5626,31 @@ class DecomposeAtenAdaptiveAvgPool2dOp
                                  : rewriter.create<Torch::ConstantIntOp>(
                                        loc, rewriter.getI64IntegerAttr(
                                                 inputShape[rank - 2 + i])));
+        strides.push_back(constantOne);
```
Comment (zjgarvey) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:5629]: Now that strides == kernelSizeValues below, this seems to be the only place where those might differ, correct? Is it necessary to have stride = 1 in this case, or can you get rid of the separate stride prim list altogether and just pass the kernelSizeInts to the AtenAvgPool op?

Regardless, if you'd prefer to keep them separate for readability, I think that's fine. too.

**Review by zjgarvey — APPROVED** [review]
Thanks for making the requested changes. I have one other small suggestion after looking at the code again. Let me know if you'd like to merge it as-is, or if you make other changes, feel free to request another review.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 5 above)
Comment (Xinyu302) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: It is not necessary to discuss the case of unitOutputSize=1. So I eliminated the branch and modified the test case. What do you think about my latest change? By the way, thank you very much for your review.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 6]
```diff
@@ -5593,65 +5591,37 @@ class DecomposeAtenAdaptiveAvgPool2dOp
...
+                                         "output size");
       }
+      Value stride = rewriter.create<AtenFloordivIntOp>(
+          loc, inputHW[i], outputShapeSizesTorchInt[i]);
+      Value kernelSizeValue = stride;
+      kernelSize.push_back(kernelSizeValue);
+      strides.push_back(stride);
     }
 
     Value kernelSizeList = rewriter.create<PrimListConstructOp>(
```
[the remark this answers is not in the document]
Comment (Xinyu302) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:5617]: modified the code based on your comments

**Review by zjgarvey — APPROVED** [review]
LGTM

**Comment by zjgarvey** [comment]
@renxida I don't have write access, so would you mind taking a glance, resolving conversations, and merging this if @Xinyu302 is satisfied?

**Review by qingyunqu — APPROVED** [review]
LGTM

**Pull request merged**